### DISCUSSION TITLE
Question about the runtime error: Segmentation fault (core dumped)

### DISCUSSION THREAD

**Discussion by LuvMangorange** [mhm-ufz/mHM#25] [Q&A] [answered]
Hey guys
Where did this error come from? 
![image](https://user-images.githubusercontent.com/34867664/123199802-2269f980-d4e2-11eb-94d6-ec4f33ece22c.png)
Thanks for help!

**Comment by pallavhydro** [comment]
Hi 

Segmentation faults themselves don't convey much message. But the timing they appear during the program gives some clues. 

Could you please share the full runtime screen of the error?

Pallav

**Reply by LuvMangorange** [reply]
Hi
Of course！Picture 1 is from WSL system, picture 2 is from Ubuntu system.
I guess the error may come from the morphological data, but I have not found a solution yet.
The main production steps of the files are as follows:
1. recondition dem：generate AgreeDem by river.shp
2. generate slope & aspect 
3. fill sink
4. generate fdir & facc
Thanks for help!

Picture 1:
![image](https://user-images.githubusercontent.com/34867664/123363788-d3859800-d5a5-11eb-859d-10fbf83717d3.png)
Picture 2:
![image](https://user-images.githubusercontent.com/34867664/123363826-e8fac200-d5a5-11eb-8bef-54013afdcf13.png)

**Reply by pallavhydro** [reply]
Hi

Thanks for the screenshots. Obviously the `ubuntu screenshot` seems to give a better runtime error output in this case.

The error `Something went wrong during L11_link_location... movedownonecell got stuck in infinite loop at cell ...` means the fdir.asc has some problematic cells that don't flow into a next cell. 

However, even before that, the runtime screen has more clues. The `Number of outlets found at Level 0:` is 0. This is the problem. Your domain doesn't have an outlet at L0 level. L0 (or level 0) is the resolution of dem, facc, fdir, ... while L11 (or level 11) is model resolution. If you compare this with the test domain of mhm, you will notice that the test domain has 1 outlet at L0. The number of outlets at L0 needs to be 1 or more, and can't be 0. In the latter case, you end up with the current error. 

Way forward: 
- `Check`: Visualise your fdir and facc. Can you see the facc (in GIS) ending up at the edge of the domain? Share screenshots if feasible.
- `Solution`: Apparently your fdir has some issues. I would have reprocessed dem. I am not sure about what happens to the dem in the step `recondition dem`. Perhaps the above check would give you some clues.

Pallav

**Reply by LuvMangorange** [reply]
Hi 
I have done it twice from the begining, but the result is still. I want to know how the model exams wheather there is a outlet in leavel 0? 
This is the screenshot of the facc and outlet:
![image](https://user-images.githubusercontent.com/34867664/123424675-be375a80-d5f3-11eb-8acc-ac91ee322e2b.png)

**Reply by pallavhydro** [reply]
Hi

mHM uses fdir to move downstream from each cell. An outlet is the condition when a cell flows to a downstream cell that is -9999 i.e. out of domain (white area in the above screenshot). 

Could you please share the facc, fdir and dem that you are currently using as input to mhm? I will take a look at them.

Cheers
P

**Reply by LuvMangorange** [reply]
Hi
Thank you very much!

**Comment by LuvMangorange** [comment, marked as answer]
Hi @pallavhydro
The error has been resolved.

The error of "number of outlet at L0 is 0" can refer to my share "What happens when the resolution of L0 & L2 are the same" #26.
The error of "the movedowncell got stuck in..." was caused by the watershed. In the begining, the watershed I used was obtained from other sources not directly generated by the dem, so some error can occur at the edge of the flow direction file.

Anyway, thanks for your kind help!
